File: backend/app/workers/monitor_worker.py
import time
import httpx
import asyncio
from app.websocket.connection_manager import manager
from sqlalchemy.orm import Session
from app.database.session import SessionLocal
from app.models.endpoint import Endpoint
from app.models.monitor_result import MonitorResult
from app.models.project import Project
from app.models.user import User
from app.services.alert_service import check_failure_threshold, create_alert
from app.services.notification_service import send_email_alert


async def check_endpoint(endpoint_id):
    """
    Check a single endpoint, record result, send alert or recovery if needed.
    """

    db: Session = SessionLocal()
    # Refetch fresh endpoint from DB to get updated alert_active state
    endpoint = db.query(Endpoint).filter(Endpoint.id == endpoint_id).first()
    if not endpoint or not endpoint.is_active:
        db.close()
        return

    start = time.time()
    success = False
    status_code = None

    try:
        async with httpx.AsyncClient(timeout=endpoint.timeout_seconds) as client:
            response = await client.request(endpoint.method, endpoint.url)
            status_code = response.status_code
            if status_code < 500:
                success = True
    except Exception:
        success = False

    response_time = (time.time() - start) * 1000

    # Record monitoring result
    result = MonitorResult(
        endpoint_id=endpoint.id,
        status_code=status_code,
        response_time=response_time,
        success=success
    )
    db.add(result)
    db.commit()


    await manager.broadcast({
        "endpoint_id": str(endpoint.id),
        "endpoint_name": endpoint.name,
        "status_code": status_code,
        "response_time": response_time,
        "success": success
    })

    # -----------------------------
    # ALERT LOGIC
    # -----------------------------

    # Fetch project and user once
    project = db.query(Project).filter(Project.id == endpoint.project_id).first()
    user = None
    if project:
        user = db.query(User).filter(User.id == project.user_id).first()

    # Case 1 — failure
    if not success:
        threshold_reached = check_failure_threshold(db, endpoint.id)
        if threshold_reached and not endpoint.alert_active:
            # Create alert
            alert = create_alert(
                db,
                endpoint.id,
                f"Endpoint {endpoint.name} failed 3 consecutive checks"
            )

            # Send email alert
            if user:
                try:
                    send_email_alert(user.email, "API Sentinel Alert", alert.message)
                except Exception as e:
                    print("Email sending failed:", e)

            # Mark alert active
            endpoint.alert_active = True
            db.commit()

    # Case 2 — recovery
    else:
        if endpoint.alert_active:
            # Send recovery email
            if user:
                try:
                    send_email_alert(
                        user.email,
                        "API Sentinel Recovery",
                        f"Endpoint {endpoint.name} is back UP"
                    )
                except Exception as e:
                    print("Recovery email failed:", e)

            # Reset alert state
            endpoint.alert_active = False
            db.commit()

    db.close()


async def run_monitoring():
    db = SessionLocal()
    # Fetch all active endpoints
    endpoints = db.query(Endpoint).filter(Endpoint.is_active == True).all()
    db.close()

    tasks = [check_endpoint(e.id) for e in endpoints]
    await asyncio.gather(*tasks)

File: backend/src/workers/monitor.worker.js
const axios = require('axios');
const prisma = require('../database/client');
const { manager } = require('../websocket/connection_manager');
const { checkFailureThreshold, createAlert } = require('../services/alert.service');
const { sendEmailAlert } = require('../services/notification.service');

async function checkEndpoint(endpointId) {
  const endpoint = await prisma.endpoint.findUnique({ where: { id: endpointId } });
  if (!endpoint || !endpoint.is_active) return;

  const start = Date.now();
  let success = false;
  let status_code = null;

  try {
    const response = await axios({
      method: endpoint.method,
      url: endpoint.url,
      timeout: endpoint.timeout_seconds * 1000, // axios takes ms, not seconds
    });
    status_code = response.status;
    success = status_code < 500;
  } catch (err) {
    status_code = err.response?.status || null;
    success = false;
  }

  const response_time = Date.now() - start;

  await prisma.monitorResult.create({
    data: { endpoint_id: endpoint.id, status_code, response_time, success },
  });

  manager.broadcast({
    endpoint_id: endpoint.id,
    endpoint_name: endpoint.name,
    status_code,
    response_time,
    success,
  });

  // Fetch project → user for email
  const project = await prisma.project.findUnique({ where: { id: endpoint.project_id } });
  const user = project
    ? await prisma.user.findUnique({ where: { id: project.user_id } })
    : null;

  if (!success) {
    const threshold_reached = await checkFailureThreshold(endpoint.id);
    if (threshold_reached && !endpoint.alert_active) {
      const alert = await createAlert(
        endpoint.id,
        `Endpoint ${endpoint.name} failed 3 consecutive checks`
      );
      if (user) {
        try { await sendEmailAlert(user.email, 'API Sentinel Alert', alert.message); }
        catch (e) { console.error('Email failed:', e.message); }
      }
      await prisma.endpoint.update({
        where: { id: endpoint.id },
        data: { alert_active: true },
      });
    }
  } else {
    if (endpoint.alert_active) {
      if (user) {
        try {
          await sendEmailAlert(
            user.email,
            'API Sentinel Recovery',
            `Endpoint ${endpoint.name} is back UP`
          );
        } catch (e) { console.error('Recovery email failed:', e.message); }
      }
      await prisma.endpoint.update({
        where: { id: endpoint.id },
        data: { alert_active: false },
      });
    }
  }
}

async function runMonitoring() {
  const endpoints = await prisma.endpoint.findMany({ where: { is_active: true } });
  await Promise.all(endpoints.map(e => checkEndpoint(e.id)));
}

module.exports = { runMonitoring };
